### PR TITLE
SWATCH-1521 & SWATCH-1775: Configure Rhel for EUS For Swatch

### PR DESCRIPTION
**Pull request by ksynvrit**
Jira issue: [SWATCH-1521](https://issues.redhat.com/browse/SWATCH-1521) & Jira issue: [SWATCH-1775](https://issues.redhat.com/browse/SWATCH-1775)
## Description

This is to configure `Rhel for x86 EUS` for traditional subscriptions so it may be queried . Effectively, anywhere you can query/interact with `RHEL for x86`. 
Since the eng ids are same i.e `70` for `Rhel for x86 EUS Payg` and `non-payg` we want to make sure the correct product gets pulled when we do hourly tally and nightly tally. Hence there are two instructions one for nightly tally shouldn't pull payg products and other hourly tally should pull payg products.

## Testing instructions for nightly tally.

1. Add host to Insight DB:
```
INSERT INTO public.hosts (id, account, display_name, created_on, modified_on, facts, tags, canonical_facts, system_profile_facts, ansible_host, stale_timestamp, reporter, per_reporter_staleness, org_id) VALUES ('e5406d59-5cf3-4e83-abf8-84c149f8b6ba', 'account123', 'd125f119-0c98-43f2-a5de-ccf2e8e1ab55', '1993-03-26 00:00:00.000000 +00:00', '1993-03-26 00:00:00.000000 +00:00', '{"rhsm": {"orgId": "org123", "IS_VIRTUAL": null, "VM_HOST_UUID": null}}', null, '{"insights_id": "d125f119-0c98-43f2-a5de-ccf2e8e1ab55", "subscription_manager_id": "ef5a9896-242c-44a4-908d-9e6a1ffb9df6"}', '{"host_type": null, "cloud_provider": null, "is_marketplace": false, "cores_per_socket": 4, "number_of_sockets": 1, "installed_products": [{"id": "70"}], "infrastructure_type": "PHYSICAL"}', null, '2030-01-01 00:00:00.000000 +00:00', 'rhsm-conduit', '{}', 'org123');

```
2. Run nightly Tally using command below:
```
http PUT ":8000/api/rhsm-subscriptions/v1/internal/rpc/tally/snapshots/org123" \
Origin:console.redhat.com \
x-rh-swatch-synchronous-request:true \
x-rh-swatch-psk:placeholder
```

### Verification

Observe that Tables in `TallySnapsots`, `Host_Tally_Bucket` & `Instance_measurements` has the configured value: `RHEL for x86 EUS` and it configured MetricId : `SOCKETS`

## Testing instructions for Hourly tally- The steps will be same as [PR SWATCH-1690](https://github.com/RedHatInsights/rhsm-subscriptions/pull/2560)
1. Insert an event manually in Events table. You can follow some steps in SWATCH-1547 to create events using Kafka.
`INSERT INTO public.events (event_id, account_number, timestamp, data, event_type, event_source, instance_id, org_id,  metering_batch_id, record_date) VALUES ('16893602-d1f9-414d-8bf6-a90154e844bf', '10001', '2023-09-01 08:00:00.000000 +00:00', '{"sla": "Self-Support", "role": "Red Hat Enterprise Linux Workstation", "usage": "Disaster Recovery", "org_id": "1234567", "event_id": "16893602-d1f9-414d-8bf6-a90154e844bf", "timestamp": "2023-09-01T08:00:00Z", "event_type": "Snapshot2", "expiration": "2023-08-01T08:00:00Z", "instance_id": "i-55555558", "product_ids": ["479", "70"], "event_source": "cost-management", "measurements": [{"uom": "vCPUs", "value": 3.0}], "service_type": "RHEL System", "hardware_type": "Cloud", "account_number": "10001", "cloud_provider": "AWS", "billing_provider": "aws", "billing_account_id": "9999999999999"}', 'Snapshot2', 'cost-management', 'i-55555558', '1234567', null, '2023-09-01 08:00:00.000000 +00:00');`

2. Insert a record manually in `Contracts` table and `Contract_metrics` table. If you don't do this then you won't be able to test sending usage to AWS. You also need a Subscription and Offering that are not available. In prod and stage a contract should come from API gateway in UMB message.
- `INSERT INTO public.contracts (uuid, subscription_number, last_updated, start_date, end_date, org_id, sku, billing_provider, billing_account_id, product_id, vendor_product_code) VALUES ('a3a5e8b7-6260-48f4-a9eb-63a584763ce8', '12914276', '2023-05-10 18:20:07.717275 +00:00', '2023-05-10 18:20:07.717275 +00:00', null, '1234567', 'MW02393', 'aws', '9999999999999', 'rhel-for-x86-eus-payg', 'bcwuzc8ww10vvxfair55mliy8');`

- `INSERT INTO public.contract_metrics (contract_uuid, metric_id, metric_value) VALUES ('a3a5e8b7-6260-48f4-a9eb-63a584763ce8', 'VCPUS', 10);`

### Steps

1.Run in separate terminals.
**Tally:**
`RHSM_SUBSCRIPTIONS_ENABLE_SYNCHRONOUS_OPERATIONS=true DEV_MODE=true PROM_URL=http://localhost:8082/api/metrics/v1/telemeter/api/v1 SWATCH_CONTRACTS_INTERNAL_SERVICE_URL=http://localhost:8882 SERVER_PORT=8001 ./gradlew clean :bootRun
`

**Swatch-contract:**
`SERVER_PORT=8882 ./gradlew  :swatch-contracts:quarkusDev
`

**Swatch-aws-producer:**
`SWATCH_INTERNAL_SUBSCRIPTION_ENDPOINT=http://localhost:8001/api/rhsm-subscriptions/v1 ./gradlew swatch-producer-aws:quarkusDev
`

**Execute hourly tally:**
`http POST ":8001/api/rhsm-subscriptions/v1/internal/tally/hourly?org=1234567&start=2023-09-01T08:00Z&end=2023-09-01T09:00Z" x-rh-swatch-psk:placeholder
`
### Verification

1. Logs:
- Successfully creates usage
```
2023-09-25 11:53:20,547 [thread=swatch-producer-billing-0-C-1] [INFO ] [org.candlepin.subscriptions.tally.billing.BillableUsageController] - Finished producing monthly billable for orgId=1234567 productId=rhel-for-x86-eus-payg uom=VCPUS provider=aws, snapshotDate=2023-09-01T08:00Z
2023-09-25 11:53:20,547 [thread=swatch-producer-billing-0-C-1] [INFO ] [org.candlepin.subscriptions.tally.billing.BillingProducer] - Sending billable usage org.candlepin.subscriptions.json.BillableUsage@7b5b059a[accountNumber=10001,orgId=1234567,id=d9e5484a-e52e-409b-bf0e-c0f7dec2837a,billingProvider=aws,billingAccountId=9999999999999,snapshotDate=2023-09-01T08:00Z,productId=rhel-for-x86-eus-payg,sla=Self-Support,usage=Disaster Recovery,uom=VCPUS,value=0.0,billingFactor=1.0,vendorProductCode=<null>] to topic platform.rhsm-subscriptions.billable-usage
```

- Since this is not for Rh-marketplace:
`2023-09-25 11:53:20,549 [thread=swatch-producer-rh-marketplace-0-C-1] [WARN ] [org.candlepin.subscriptions.rhmarketplace.RhMarketplacePayloadMapper] - Skipping invalid billable usage org.candlepin.subscriptions.json.BillableUsage@205146a0[accountNumber=10001,orgId=1234567,id=d9e5484a-e52e-409b-bf0e-c0f7dec2837a,billingProvider=aws,billingAccountId=9999999999999,snapshotDate=2023-09-01T08:00Z,productId=rhel-for-x86-eus-payg,sla=Self-Support,usage=Disaster Recovery,uom=VCPUS,value=0.0,billingFactor=1.0,vendorProductCode=<null>]`

- For AWS we have log messages as debug in code but, if we change the log level to DEBUG you should see other logs where it attempts to send usage but, fails because of no subscription and offering.

- No subscription and offering so will see this message.
`2023-09-25 11:53:20,930 [thread=http-nio-8001-exec-6] [ERROR] [org.candlepin.subscriptions.subscription.SubscriptionSyncController] self- No subscription found for account 10001 with criteria DbReportCriteria(accountNumber=null, orgId=1234567, productTag=null, productNames=[RHEL EUS COST PLACEHOLDER], productId=null, serviceLevel=SELF_SUPPORT, usage=_ANY, billingProvider=AWS, billingAccountId=9999999999999, payg=true, beginning=2023-09-01T07:00Z, ending=2023-09-01T08:00Z, metricId=null, hypervisorReportCategory=null)`

### CODE REVIEW

**Comment by kartikshahc** [comment]
@ksynvrit this PR should go with SWATCH-1775. Also we should make sure it doesn't throw an Exception since we have same engIds in two different product-configs.

File: swatch-product-configuration/src/main/resources/subscription_configs/RHEL/RHEL_for_x86_EUS.yaml
Comment (lindseyburnett) [inline swatch-product-configuration/src/main/resources/subscription_configs/RHEL/RHEL_for_x86_EUS.yaml:10]: We should use kebab-style for our tag names to make it more clear that this is more of an ID than a user readable string

File: swatch-product-configuration/src/main/resources/subscription_configs/RHEL/RHEL_for_x86_EUS.yaml
Comment (lindseyburnett) [inline swatch-product-configuration/src/main/resources/subscription_configs/RHEL/RHEL_for_x86_EUS.yaml:7]: Can you add a TODO comment with the JIRA number where this section becomes uncommented?

**Review by lindseyburnett — CHANGES_REQUESTED** [review]

**Review by lindseyburnett — APPROVED** [review]

**Comment by kartikshahc** [comment]
@ksynvrit I see errors after main merged in this branch:
2023-09-28 15:02:25,225 [thread=rhsm-subscriptions-task-processor-0-C-1] [ERROR] [org.candlepin.subscriptions.tally.TallySnapshotController] - Error collecting existing usage snapshots 
java.lang.IllegalArgumentException: expected one element but was: <SubscriptionDefinition(platform=RHEL, id=rhel-for-x86-eus-payg, parentSubscription=null, includedSubscriptions=[rhel-for-x86], variants=[Variant(tag=rhel-for-x86-eus-payg, roles=[], engineeringIds=[70], productNames=[RHEL EUS COST PLACEHOLDER])], serviceType=RHEL System, metrics=[Metric(id=vCPUs, rhmMetricId=null, awsDimension=vCPU_Hour, prometheus=null, billingFactor=null)], defaults=Defaults(variant=rhel-for-x86-eus-payg, sla=PREMIUM, usage=PRODUCTION), contractEnabled=true), SubscriptionDefinition(platform=RHEL, id=rhel-for-x86-eus, parentSubscription=null, includedSubscriptions=[], variants=[Variant(tag=rhel-for-x86-eus, roles=[], engineeringIds=[70], productNames=[])], serviceType=null, metrics=[Metric(id=Sockets, rhmMetricId=null, awsDimension=null, prometheus=null, billingFactor=null)], defaults=Defaults(variant=rhel-for-x86-eus, sla=null, usage=null), contractEnabled=false)>
        at com.google.common.collect.MoreCollectors$ToOptionalState.multiples(MoreCollectors.java:115)
        at com.google.common.collect.MoreCollectors$ToOptionalState.getOptional(MoreCollectors.java:157)
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:686)
        at com.redhat.swatch.configuration.registry.SubscriptionDefinition.lookupSubscriptionByEngId(SubscriptionDefinition.java:190)

**Comment by kartikshahc** [comment]
> @ksynvrit I see errors after main merged in this branch: 2023-09-28 15:02:25,225 [thread=rhsm-subscriptions-task-processor-0-C-1] [ERROR] [org.candlepin.subscriptions.tally.TallySnapshotController] - Error collecting existing usage snapshots java.lang.IllegalArgumentException: expected one element but was: <SubscriptionDefinition(platform=RHEL, id=rhel-for-x86-eus-payg, parentSubscription=null, includedSubscriptions=[rhel-for-x86], variants=[Variant(tag=rhel-for-x86-eus-payg, roles=[], engineeringIds=[70], productNames=[RHEL EUS COST PLACEHOLDER])], serviceType=RHEL System, metrics=[Metric(id=vCPUs, rhmMetricId=null, awsDimension=vCPU_Hour, prometheus=null, billingFactor=null)], defaults=Defaults(variant=rhel-for-x86-eus-payg, sla=PREMIUM, usage=PRODUCTION), contractEnabled=true), SubscriptionDefinition(platform=RHEL, id=rhel-for-x86-eus, parentSubscription=null, includedSubscriptions=[], variants=[Variant(tag=rhel-for-x86-eus, roles=[], engineeringIds=[70], productNames=[])], serviceType=null, metrics=[Metric(id=Sockets, rhmMetricId=null, awsDimension=null, prometheus=null, billingFactor=null)], defaults=Defaults(variant=rhel-for-x86-eus, sla=null, usage=null), contractEnabled=false)> at com.google.common.collect.MoreCollectors$ToOptionalState.multiples(MoreCollectors.java:115) at com.google.common.collect.MoreCollectors$ToOptionalState.getOptional(MoreCollectors.java:157) at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:686) at com.redhat.swatch.configuration.registry.SubscriptionDefinition.lookupSubscriptionByEngId(SubscriptionDefinition.java:190)

Please let me know if you need help in this. Also there were some failing tests.

**Review by kartikshahc — CHANGES_REQUESTED** [review]
A failing test and nightly tally fails due to multiple product ids.

**Review by kartikshahc — APPROVED** [review]
Since I have also made changes in code I would like to get a review from other reviewer.

**Review by lindseyburnett — APPROVED** [review]

**Review by TrayvonMcKnight — APPROVED** [review]
https://drive.google.com/file/d/1WPztI669XYPf05nIZ6e9LMIs0rrHJ93G/view?usp=drive_link

**Pull request merged**
